### DISCUSSION TITLE
Ticket Activity - Add to log

### DISCUSSION THREAD

**Discussion by LeeTV1** [cerb/cerb-release#62] [Q&A] [answered]
Hello:

I want to be able to show in the log when a user opens a ticket (displayed it) on their screen. 

I feel this will help with accountability and tracking. 

Where do I get started?

Any help would be much appreciated. 

Lee

**Comment by blancsw** [comment]
Hi Lee,
You want to log something in the "Automation Logs" window ?

setup >> Developers >> Automation Logs

**Comment by jstanden** [comment, marked as answer]
Hi Lee!

The recent [10.4](https://cerb.ai/releases/10.4/) update modifies the `record.viewed` automation event so it triggers on both profile pages and card popups.

There's an example of creating custom activity log entries here:
https://cerb.ai/docs/api/records/activity-log/#create-an-activity-log-entry

You can use the [record.create:](https://cerb.ai/docs/automations/commands/record.create/) command in an automation to create [log records](https://cerb.ai/docs/records/types/activity_log/).

You'd probably also want your automation to have a cooldown so it doesn't log every record view from the same worker. It could log a distinct worker+record per day. You could search log entries before creating one, but keep in mind this will have a performance impact.

[Metrics](https://cerb.ai/docs/metrics/) are another option. They're efficient, but you'll end up with a lot of data based on the cardinality of `workers * tickets`. The retention of daily metrics is currently forever. That may be fine depending on your use case.

**Reply by jstanden** [reply]
If you really want to log *every* ticket view.

Import this package from **Setup >> Packages**:

```json
{
    "package": [],
    "records": [
        {
            "uid": "automation_64dabfbb6e18f",
            "_context": "automation",
            "name": "example.records.logTicketViews",
            "extension_id": "cerb.trigger.record.viewed",
            "description": "Log every ticket viewed by a worker",
            "script": "start:\r\n  outcome/notTicket:\r\n    if@bool: {{record__type is not record type ('ticket')}}\r\n    then:\r\n      return:\r\n\r\n  record.create/log:\r\n    output: new_log_entry\r\n    inputs:\r\n      record_type: activity_log\r\n      fields:\r\n        activity_point: custom.other\r\n        actor__context: worker\r\n        actor_id@int: {{worker_id}}\r\n        target__context: ticket\r\n        target_id@int: {{record_id}}\r\n        params:\r\n          message@raw: {{worker}} viewed ticket {{target}}\r\n          variables:\r\n            worker: {{worker__label}}\r\n            target: {{record__label}}\r\n          urls:\r\n            worker: ctx://worker:{{worker_id}}\r\n            target: ctx://ticket:{{record_id}}",
            "policy_kata": "commands:\r\n  record.create:\r\n    deny/type@bool: {{inputs.record_type is not record type ('activity_log')}}\r\n    allow@bool: yes",
            "created_at": 1692057531,
            "updated_at": 1692057531
        }
    ]
}
```

Then edit **Search >> Automation Events >> record.viewed** and bind the automation to the event:

```yaml
automation/ticketViewed:
  uri: cerb:automation:example.records.logTicketViews
  disabled@bool: {{record__type is not record type ('ticket')}}
```

You'll get endless activity log entries like:

<img width="599" alt="CleanShot 2023-08-14 at 17 00 50@2x" src="https://github.com/cerb/cerb-release/assets/63288/0acfc4d2-6a3b-46d2-aa4f-5d1b9dcaad9e">

**Comment by LeeTV1** [comment]
Thank you for sending this over. I received this error when I tried to add this to Setup > Packages > Import

xError validating record (automation_64dabfbb6e18f): 'Trigger' (cerb.trigger.record.viewed) is not a valid extension ID on (cerb.automation.trigger).

**Reply by Hildy** [reply]
You're on 10.3.10 right now, Lee.  As Jeff noted, this event was split in 10.4 (and so names changed).  You'll want to update to the latest before importing that package.
